File: python/modules/UseRandom.py
from icecube.icetray import *
from icecube import dataclasses

class UseRandom(I3Module):
    """
    This module class uses an I3RandomService to get doubles and put them in
    the frame.  Used to illustrate that the randomservice can be pure python
    (e.g. ConstantService) or wrapped cpp (I3GSLRandomService).

    *Parameters*:

    * **I3RandomService** - a random service instance
    * **PutWhere** - where to put the constructed doubles
    
    """
    def __init__(self, context):
        I3Module.__init__(self, context)
        self.AddParameter("I3RandomService", "the service", None)
        self.AddParameter("PutWhere", "where the doubles go", None)

    def Configure(self):
        self.rs = self.GetParameter("I3RandomService")
        self.where = self.GetParameter("PutWhere")

    def Physics(self, frame):

        rnd = self.rs.Gaus(0,1)
        d = dataclasses.I3Double(rnd)
        frame.Put(self.where, d)

        self.PushFrame(frame)


        

File: private/modules/UseRandom.cxx
#include <icetray/I3Frame.h>
#include <icetray/I3Logging.h>
#include <icetray/I3Module.h>

#include <dataclasses/I3Time.h>
#include <dataclasses/I3Double.h>
#include <phys-services/I3RandomService.h>

class UseRandom : public I3Module
{
  I3RandomServicePtr rs;
  std::string key;

 public:

  UseRandom(const I3Context& ctx) : I3Module(ctx)
  {
    AddParameter("I3RandomService",
		 "my random service",
		 rs);

    AddParameter("PutWhere",
		 "where the doubles go",
		 key);

    AddOutBox("OutBox");
  }

  void Configure()
  {
    GetParameter("I3RandomService", rs);
    log_debug("rndserv is at %p", rs.get());
    GetParameter("PutWhere", key);
  }

  void Physics(I3FramePtr frame)
  {
    log_debug("rndserv is at %p", rs.get());
    double d = rs->Gaus(0, 1);
    I3DoublePtr dp(new I3Double(d));
    frame->Put(key, dp);
    PushFrame(frame);
  }
};

I3_MODULE(UseRandom);

